src/atlas/grid/detail/partitioner/CubedSpherePartitioner.cc

```diff
@@ -0,0 +1,248 @@
+
+            offsetPerTile[proc1D] = proc1D == 0 ? 0 :
+                                                  std::accumulate(ngpbt.begin() + jt, ngpbt.begin() + j, 0) / maxDim2D[t];
+
+        }
+        offsetPerTile[nproc] = maxDim1D[t];
+        offset.push_back(offsetPerTile);
+    }
+
+    return offset;
+}
+
+
+}
+
+CubedSpherePartitioner::CubedSpherePartitioner() : Partitioner() {}
+
+CubedSpherePartitioner::CubedSpherePartitioner( int N ) : Partitioner( N ), regular_{true} {}
+
+CubedSpherePartitioner::CubedSpherePartitioner( int N, const eckit::Parametrisation& config ) : Partitioner( N ), regular_{false} {
+    config.get( "starting rank on tile", globalProcStartPE_);
+    config.get( "final rank on tile", globalProcEndPE_);
+    config.get( "nprocx", nprocx_);
+    config.get( "nprocy", nprocy_);
+}
+
+CubedSpherePartitioner::CubedSpherePartitioner( const int N, const std::vector<int> & globalProcStartPE,
+                                                const std::vector<int> & globalProcEndPE,
+                                                const std::vector<int> & nprocx,
+                                                const std::vector<int> & nprocy)
+    : Partitioner( N ), globalProcStartPE_{globalProcStartPE},  globalProcEndPE_{globalProcEndPE},
+      nprocx_{nprocx},  nprocy_{nprocy}, regular_{false}   {}
+
+
+CubedSpherePartitioner::CubedSphere CubedSpherePartitioner::cubedsphere( const Grid& grid ) const {
+    // grid dimensions
+    const CubedSphereGrid cg( grid );
+    if ( !cg ) {
+        throw_Exception( "CubedSphere Partitioner only works for cubed sphere grids.", Here() );
+    }
+
+    CubedSphere cb;
+
+    for (std::size_t t = 0; t < 6; ++t) {
+        cb.nx[t] = cg.N();
+        cb.ny[t] = cg.N();
+    }
+
+    atlas::idx_t nparts = nb_partitions();
+
+    if (regular_) {
+        // share PEs around tiles
+        // minRanksPerTile
+
+        idx_t ranksPerTile = nparts/6;
+
+        idx_t reminder = nparts - 6 * ranksPerTile;
+
+        for (std::size_t t = 0; t < 6; ++t) {
+            cb.nproc[t] = ranksPerTile;
+        }
+
+        // round-robin;
+        std::size_t t{0};
+        while (reminder > 0) {
+            if (t == 6) t=0;
+            cb.nproc[t] += 1;
+            t += 1;
+            reminder -= 1;
+        }
+
+        // now need to specify nprocx and nprocy.
+        // nproc is 0 for the tile we use default nprocx and nprocy = 1
+
+        // if we can square-root nproc and get an integer
+        // we use that for nprocx and nprocy
+        // otherwise we split just in nprocx and keep nprocy =1;
+
+        for (size_t t = 0; t < 6; ++t) {
+            if (cb.nproc[t] > 0) {
+                double sq = std::sqrt(static_cast<double>(cb.nproc[t]));
+                if (isNearInt(sq)) {
+                    cb.nprocx[t] = static_cast<idx_t>(floor(sq+0.5));
+                    cb.nprocy[t] = static_cast<idx_t>(floor(sq+0.5));
```

Comment:
Nitpicking, but `std::round` might be a bit more idiomatic for C++ here.

Reply:
Done!